### DISCUSSION TITLE
Constructor-only called methods

### DISCUSSION THREAD

**Discussion by SmartK8** [dotnet/csharplang#4704] [Language Ideas]
At this point, we're forced to initialize read-only fields in a constructor only (or a single field via initializer). This is fine when the initialization code is straightforward, but when it reaches a certain complexity, it impedes readability. It would be great if we could call some kind of marked methods (a compilation trick) to make code more readable. These methods would be only allowed to be called from the constructors. It would just put all the called methods inside the constructor upon compilation in the call order. Example:

Now this is mess:

```
public struct/record/class MessyConstructor
{
    // many read-only fields

    public readonly Int32 A;
    public readonly Int32 B;
    ...

    public MessyConstructor(Byte[] data)
    {
        ...
        A complex method - all the fields must be set just here.
        Many times you can return just field values from the methods.
        But sometimes one method produces many results.
        You can of course return tuples, out variables, etc. 
        It works for simple methods but reduces readability for the complex cases.
        It can especially become mess if you modify fields on multiple places.
        ...
    }
}
```

After newly introduced trick (access modifier on **init** methods is replaced):

```
public struct/record/class NeatConstructor
{
    // many read-only fields

    public readonly Int32 A;
    public readonly Int32 B;
    ...

    init void PerformPartOfProcessing1()
    {
        // complex calculation setting many fields (A, B..)
    }

    init void PerformPartOfProcessing2()
    {
        // complex calculation setting many fields (other fields)
    }

    public NeatConstructor(Byte[] data)
    {
        init void PerformPartOfProcessing3()
        {
            // another complex calculation setting many fields (other fields)
        }

        PerformPartOfProcessing1();
        PerformPartOfProcessing2();
        PerformPartOfProcessing3();
    }

    init static void PerformPartOfProcessing4() {} // fails to compile
    public Int32 SomeField = PerformPartOfProcessing1(); // fails to compile
    public Int32 SomeProperty => PerformPartOfProcessing1(); // fails to compile

    public void SomeMethod()
    {
        PerformPartOfProcessing1(); // fails to compile
    }
}
```

After compilation replacement (compilation fails removed):

```
public struct/record/class NeatConstructor
{
    // many read-only fields

    public readonly Int32 A;
    public readonly Int32 B;
    ...

    public NeatConstructor(Byte[] data)
    {
        // complex calculation setting many fields (A, B..)
        // complex calculation setting many fields (other fields)
        // another complex calculation setting many fields (other fields)
    }
}
```

**Comment by TonyValenti** [comment]
Love this idea!

I read your title and thought "There should be 'init' methods!" then read your post and saw that was the exact conclusion you came to!

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-06-15.md#init-methods

These design notes indicate that the team is interested but that there are some challenges that would have to be addressed first.

Note that you can pass a `readonly` field as a `ref` parameter so you can break the initialization up into multiple methods like this:

```cs
public class Foo {
    private readonly int a;
    private readonly int b;
    private readonly int c;
    private readonly int d;

    public Foo() {
        PerformSomeInitHere(ref a, ref b);
        AndSomeMoreInitHere(ref c, ref d);
    }

    private void PerformSomeInitHere(ref int a, ref int b) {
        // calc and init a and b here
    }

    private void AndSomeMoreInitHere(ref int c, ref int d) {
        // calc and init c and d here
    }
}
```

**Reply by SmartK8** [reply]
I'm glad it was considered and I like the conclusion. Yeah, using **ref** is a bit cleaner then **out**, but still pain if we're talking about maintaining complex scenarios.

**Reply by bernd5** [reply]
Nice trick - it will just not work for structs (error CS0170: Use of possibly unassigned field)

**Comment by koszeggy** [comment]
I also like the idea. When applied to virtual methods, it could be a good tool also to avoid the false alarms for [CA2214](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2214) where it is known that the class is not initialized yet as this method _is_ the initializer.

**Comment by TahirAhmadov** [comment]
This is a good idea but I think it's important to remember the flow analysis aspect of it. If an `init` method is called, which assigns a field, that field has to be marked as initialized for purposes of NRTs.

**Reply by 333fred** [reply]
> If an init method is called, which assigns a field, that field has to be marked as initialized for purposes of NRTs.

This is unlikely to ever happen. `MemberNotNull` or similar would be needed for that purpose.

**Comment by gfraiteur** [comment]
We (Metalama) would also welcome this feature.

An init-only method:
* would have private accessibility
* would be allowed to be called only from the constructor or an init accessor
* would be allowed to set read-only fields and init-only properties or indexers

Without this feature, it is quite difficult to generate code for a refactoring that needs to wrap the init accessor of an indexer (`this[...]`) with some idiomatic C# construct.  For properties, one can always wrap a property into another property of the same type but a different name. But for indexers, you cannot change the name, so wrapping an indexer involves, oddly enough, changing the type of the index argument. Say we have an indexer `this[int]`. If we want to wrap it, we need to "rename" the original indexer `this[Wrapped<int>]` and create a new indexer with the original type. This is quite a frequent kind of code rewriting, but it's aweful when we combine indexers and init.

**Comment by pelletburner** [comment]
I support this.

Makes implementation of complex initialization much cleaner.
